### DISCUSSION TITLE
Unexpected Spike in train/kobj_loss at Epoch 90 While Training YOLOv8m Pose Estimation Model

### DISCUSSION THREAD

**Discussion by UranLajciSolaborate** [ultralytics/ultralytics#16303] [Q&A] [answered]
I am observing a peculiar behavior in the train/kobj_loss (Keypoint Objectness Loss) during training. Specifically, around epoch 90, there's a sudden spike in the kobj_loss, which increases from 0.149 to 0.196, before stabilizing and decreasing again after epoch 90 (see attached image for reference).
![image](https://github.com/user-attachments/assets/2f680da2-45f2-46fe-b0d4-816daefef4bf)
My questions are:

    What could be causing this sudden spike in kobj_loss?
    Are there any typical adjustments or investigations I should perform to address this issue?
    Is this behavior expected in any case, such as learning rate changes or data-related factors?

I’d appreciate any insights or suggestions on what might be causing this issue and how to address it.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @UranLajciSolaborate, thank you for reaching out and for using Ultralytics 🚀! Your curiosity about the spike in `train/kobj_loss` is understandable, and we'll certainly work through it.

To assist further, if this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us analyze it.

🔍 If you're seeking advice on training dynamics, such as sudden loss spikes, consider these steps:

1. Ensure your training setup aligns with our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).
2. Check if any scheduled learning rate changes are impacting your results.
3. Verify your dataset consistency to rule out potential data anomalies.

For real-time troubleshooting, join us on [Discord](https://ultralytics.com/discord) 🎧 or explore detailed threads on our [Discourse](https://community.ultralytics.com).

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLOv8 can be run in any of the following up-to-date verified environments:

- **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon**: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are currently passing.

This is an automated response, but rest assured, an Ultralytics engineer will assist soon. Thank you for your patience! 😊

**Comment by glenn-jocher** [comment]
@UranLajciSolaborate close mosaic kicks in 10 epochs before the end of training to help the model adapt to more real-world data it might encounter rather than the earlier training mosaics it is provided. This is causing the changes in loss you see.

**Reply by UranLajciSolaborate** [reply]
Hello @glenn-jocher. In the plot above is shown the loss of the model that has the close mosaic set to 10.

**Reply by glenn-jocher** [reply]
Hello @UranLajciSolaborate. The spike in loss around epoch 90 is likely due to the close mosaic setting, which adjusts data augmentation to stabilize training.

**Reply by UranLajciSolaborate** [reply]
What are your reasons?

**Reply by glenn-jocher** [reply]
The loss spike is clearly correlated with close mosaic since these both happen at epoch 90.

**Reply by UranLajciSolaborate** [reply]
Okey. Currently I have set close mosaic set to 10 and I am getting this spike. What do you propose to do in order to escape it. Should I make this argument 20 or some larger number or remove it completely.

**Reply by glenn-jocher** [reply, marked as answer]
You can try increasing the close mosaic value to 20 to see if it stabilizes the loss, or experiment with removing it to observe the effect on your model's performance.

**Reply by UranLajciSolaborate** [reply]
Hello @glenn-jocher thanks for the Q&A interaction.

I set the close_mosaic to 20 but the loss got worse as seen in the plot below:
![image](https://github.com/user-attachments/assets/f46c012f-106e-46df-a2a7-f317b7d7194c)

I will train my model by removing it completely to see its effects.

**Reply by UranLajciSolaborate** [reply]
By setting the close mosaic to 0 we handled this case of the loss spike, as seen in the loss plots below:
![image](https://github.com/user-attachments/assets/72df76bb-2b9e-4bc1-b8f9-d8f7a4381f7f)

**Reply by glenn-jocher** [reply]
@UranLajciSolaborate great to hear that setting close mosaic to 0 resolved the loss spike issue. If you have further questions, feel free to ask!
